File: java/events/MessageReceived.java
package events;

import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import javax.annotation.Nonnull;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import structures.Discord;
import structures.Command;
import structures.CommandInput;
import structures.MessageInput;

public class MessageReceived extends ListenerAdapter {
    protected Discord discord;

    public MessageReceived(Discord discord) {
        this.discord = discord;
    }

    @Override
    public void onMessageReceived(@Nonnull MessageReceivedEvent event) {
        var message = event.getMessage();
        var content = message.getContentRaw().trim();
        var prefix = "!";
        if (!content.startsWith(prefix)) return;
        if (content.equals(prefix)) return;

        var args = content.substring(prefix.length()).split("\\s+");
        var commandName = args[0].toLowerCase();

        try {
            var cls = discord.commands.get(commandName);
            if (cls != null) {
                var input = new MessageInput(message);
                var command = (Command) cls.getConstructor(Discord.class, CommandInput.class).newInstance(discord, input);
                command.run(args);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

    }
}

File: kotlin/bin/main/events/MessageReceived.kt
package events

import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import net.dv8tion.jda.api.entities.Message;
import structures.Discord
import structures.Command

class MessageReceived(private val discord: Discord) : ListenerAdapter() {
    override fun onMessageReceived(event: MessageReceivedEvent) {
        val message = event.message
        val content = message.contentRaw.trim()
        val prefix = "!"
        
        if (!content.startsWith(prefix) || content == prefix) return

        val args = content.substring(prefix.length).split("\\s+".toRegex()).toTypedArray()
        val commandName = args[0].lowercase()

        try {
            val cls = discord.commands[commandName]
            cls?.let {
                val command = it.getConstructor(Discord::class.java, Message::class.java).newInstance(discord, message) as Command
                command.run(args)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}
